### PR TITLE
Empty values and column-data-storage refactor

### PR DESCRIPTION
**Pull request by JorisGoosen**
Empty values should be stored in the data and per column.

To do this properly it required a lot of changes to the underlying code.
Ill make sure the documentation for the classes at least resembles what is going now ;)

Possible problems:
[X] importing from 0.17.3 and earlier (converting from emptyValuesMap in metadata.json)
Ignore: importing from 0.18.2 (broken emptyvalues (i might just ignore it))
[X] importing from 0.18.3 (breaking up emptyvalues into per workspace and column)
[X]] also making sure that we can still import scalar columns with a non-scalar value in them that is not convertible to a double (like in the testcsv tsv ods etc)
Should probably be a separate PR: ~~[] Add locale support (should still be done but testing will take time anyway)~~
[] importing a nominal/ordinal now as scalar means getting the ints we wrote to the _ints column. We should probably try to use the "value" shown in the label-editor instead! :o This will help with https://github.com/jasp-stats/jasp-issues/issues/2530 or https://github.com/jasp-stats/jasp-issues/issues/2325 ) (But most of that will probably be separate)

### CODE REVIEW

**Comment by JorisGoosen** [comment]
image

**Comment by JorisGoosen** [comment]
I started a bunch of builds under `emptyValuesRefactor` in case anybode feels like testing it out.

I think this is probably ready now?

**Comment by JorisGoosen** [comment]
Its maybe not completely done, but close enough that you guys could start reviewing it. 
And Ill also ask people to test it once the nightlies are build because im sure there are some things ive missed...

**Comment by JorisGoosen** [comment]
[macos arm](https://static.jasp-stats.org/Nightlies/macOS-arm64/JASP-nightly-emptyValuesRefactor-420190fe0eec179bf99a022f8538946eaaf3e123.dmg)
[macos x86](https://static.jasp-stats.org/Nightlies/macOS-x86_64/JASP-nightly-emptyValuesRefactor-420190fe0eec179bf99a022f8538946eaaf3e123.dmg)

Aah, windows doesnt work it seems...

**Comment by shun2wang** [comment]
the build test failed maybe also failing on windows.

**Comment by RensDofferhoff** [comment]
- Press new data
- add any string in an empty column
- No Type is detected
- Use this column in descriptives and the engine crashes

**Comment by JorisGoosen** [comment]
> * Press new data
> 
>     * add any string in an empty column
> 
>     * No Type is detected
> 
>     * Use this column in descriptives and the engine crashes

Oops! 

Edit: shouldve set columntype with the default values...
Nightlies are building with this fix for anyone curious

**Comment by JorisGoosen** [comment]
Aah moving labels also doesnt work anymore for columns where all values are convertible to double

**Comment by boutinb** [comment]
Some issues:

- The empty values are empty for a column, though the workspace has some empty values
- In Edit mode, if I change a value, and press the tab key, the change is lost
- With a nominal column, if I set a new value, it does not appear in the Label Editor.
- I have tried to add a computed column with just contNormal + 1: the engine crashes

Nice feature to see the labels for a scale column (if they are not the same as the value). It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.

File: CommonData/dataset.h
Comment (boutinb) [inline CommonData/dataset.h:94]: Why a pointer?
Also in Column.h: why makes EmptyValues a pointer member? Maybe because of the EmptyValues _parent initialization? By making a const pointer (it should be a const pointer), then it can be initialized.
I can add a commit for this.

File: CommonData/emptyvalues.cpp
Comment (boutinb) [inline CommonData/emptyvalues.cpp:71]: Why _hasEmptyValues should stay true if setEmptyValues is called with a an empty set of values?

**Comment by JorisGoosen** [comment]
> It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.

**Review by boutinb — CHANGES_REQUESTED** [review]
I have started to review it, it will take some times.... But already here are small comments.

File: CommonData/emptyvalues.cpp
Comment (boutinb) [inline CommonData/emptyvalues.cpp:75]: If it has a parent, it should have a look at its hasEmptyValues(), no?

File: CommonData/dataset.h
Comment (JorisGoosen) [reply CommonData/dataset.h:94]: > Maybe because of the EmptyValues _parent initialization

Yes

> it should be a const pointer

Ok sure

File: CommonData/emptyvalues.cpp
Comment (JorisGoosen) [reply CommonData/emptyvalues.cpp:71]: Because you maybe want to remove all emptyvalues for a column...

File: CommonData/emptyvalues.cpp
Comment (JorisGoosen) [reply CommonData/emptyvalues.cpp:75]: No, because that is always the workspace, thus always "has emptyvalues"

**Comment by boutinb** [comment]
If you set a column to use custom empty values, delete all empty values, and save the JASP file, then when reloading the JASP file, the 'Use custom values' for this column is unset. This is because this setting is not saved in the database, and you just try to find out whether a column uses custom empty values by checking whether it has some empty values.
The settings 'Use custom values' must be saved in the database.

**Comment by JorisGoosen** [comment]
Ooh yeah oops

**Comment by boutinb** [comment]
If you add a new column, you don't get the Label Editor in the Variables Window.

**Comment by boutinb** [comment]
In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.
Also if you change the label of the debinf column (in debug.csv), it crashes in column.cpp line 1234, because apparently the ColumnUtils::getDoubleValue fails.

File: CommonData/column.cpp
Comment (boutinb) [inline CommonData/column.cpp]: How can you be sure that the label does not exist yet?

**Comment by JorisGoosen** [comment]
> In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.

Uhm, why?
That would mess up recoding quite a bit...
Suppose youve got:
```
a	| a
b	| b
c	| c
```
And you want to recode it to:
```
1	| a
2	| b
3	| c
```

Then your proposed change would make that quite annoying

**Comment by shun2wang** [comment]
>In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.

well, I think so. because change a old value means "new value", so new value's label should be same value showing in cell. I guess in SPSS it's also same behavior, cells will showing value itself in this situation.

**Comment by boutinb** [comment]
> > In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.
> 
> Uhm, why? That would mess up recoding quite a bit... Suppose youve got:
> 
> ```
> a	| a
> b	| b
> c	| c
> ```
> 
> And you want to recode it to:
> 
> ```
> 1	| a
> 2	| b
> 3	| c
> ```
> 
> Then your proposed change would make that quite annoying

I think a common case would be that if the values are 1, 2, 3..., then people wants to recode to 0, 1, 2... or 10, 20, 30....
Then by recoding them, you don't see the recoded values in the data panel, because they get their old values as labels. Quite annoying I think

**Comment by JorisGoosen** [comment]
Maybe I could just replace the label if it is the same integer/double value and otherwise not.

**Comment by boutinb** [comment]
Maybe one solution would be to have an empty label if the value is an integer or a double value, in this case the displayed value in the data view will be the value, and if a value is changed (recoded), then no label is created.
For string values, it makes sense to have immediately labels.

**Comment by JorisGoosen** [comment]
Yeah but then once youve gone to a string label you cant go back anyway, so its the same thing.

**Comment by boutinb** [comment]
When several columns are selected, it would be nice to have the background color of the headers of all selected columns to be darker. Now only one column header is darker:
Screenshot 2024-03-14 at 16 50 39

**Comment by JorisGoosen** [comment]
Ive fixed the value changing when the label is the same and a number.

> darker color header selected columns

Good idea

Also there are still some things wrong with the labeleditor dissappearing once there are no more labels and then not coming back.

But at least for `debInf` you now dont get a crash.
But if you put `∞` as a missing value you can make the labeleditor dissappeart from it and it wont come back :p

**Comment by JorisGoosen** [comment]
New nightlies are building but I guess the windows one will fail again...

**Comment by boutinb** [comment]
Also when the missing values are not custom for a column, you don't see which missing values will be used. It would be nice to see the workspace empty values in grey in place of a empty list:
Screenshot 2024-03-14 at 16 57 02

**Comment by boutinb** [comment]
> Nice feature to see the labels for a scale column (if they are not the same as the value). It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.

Don't you think that would be nice also?
In fact I think I would always set the Label as displayed value when exists, and the original value in grey (if not the same as the label)

**Comment by boutinb** [comment]
If you open the Label Editor for a nominal column, and change one label: if you undo it, then is will undo many dummy things like "undo set name from 'contBinom' to 'contBinom'". Then it will really undo the label change. If you then redo it, it will crash (or sometimes add an extra label). This was working in. 0.18.3

**Comment by shun2wang** [comment]
> New nightlies are building but I guess the windows one will fail again...

I can build it local on Windows but Linux still failing.

**Comment by JorisGoosen** [comment]
> > Nice feature to see the labels for a scale column (if they are not the same as the value). It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.
> 
> Don't you think that would be nice also? 

Maybe, it might just make the interface a bit cluttered. If you need to see the values then you should switch to scalar.

> In fact I think I would always set the Label as displayed value when exists, and the original value in grey (if not the same as the label)

I disagree, because the data you get for a scalar column is the *values* and not the *labels*, while for ordinal/nominal its the *labels*.
And this is also why I think the value shouldnt be shown for ordinal/nominal.

File: CommonData/column.cpp
Comment (JorisGoosen) [reply CommonData/column.cpp]: We are in the function "replaceDoubleWithLabel", where we pass the dbls onwards from another function.
However, it makes no sense to replace "nan" in dbls with a label, because that is the value that would be written there if the originalValue is not convertible to double... 

So that wouldnt fit the data at all.
To enforce this and show any problems as soon as possible without confusing behaviour (given that `std::nan != std::nan` im not sure if it could ever be: `doubleIntIdMap.count(std::nan) > 0`)

**Comment by JorisGoosen** [comment]
@RensDofferhoff  @shun2wang I think linux should now be working?

**Comment by shun2wang** [comment]
Yes, it can be build on my local now.

**Comment by JorisGoosen** [comment]
Hmm, cut-copy-pasting with labels different from the values (see attached file, rename to `.sav`) 
[SPPS_Debug.zip](https://github.com/jasp-stats/jasp-desktop/files/14695665/SPPS_Debug.zip)

**Comment by boutinb** [comment]
One bug: in contBinom, set custom empty values, add 0 as empty value: you see the 0's as empty values with a 0 in grey at the right hand-side. But in Edit mode, if you click on one of these empty value and leave this cell the '0' in grey is removed. If you then remove 0 as empty value, the cell stays as an empty value. Apparently just by editing the cell, it has removed the value of its empty value.

File: Desktop/data/datasettablemodel.cpp
Comment (boutinb) [inline Desktop/data/datasettablemodel.cpp:28]: Quite strange to set a QueuedConnection here

**Review by boutinb — APPROVED** [review]
To review all the changes is quite hard...
But after some hours I think this code can be merged.

**Comment by RensDofferhoff** [comment]
I have not found any additional bugs right now.
Might be good to merge and get the related nightlies out for wider testing if you are happy with it @JorisGoosen .

**Review by RensDofferhoff — APPROVED** [review]

File: Desktop/data/undostack.cpp
Comment (JorisGoosen) [inline Desktop/data/undostack.cpp]: But arent we gonna miss errors now...

Shouldnt this be an `assert`

Comment (JorisGoosen) [reply Desktop/data/undostack.cpp]: In fact, I dont like this busy thing. Why do we need this?

**Comment by boutinb** [comment]
Oh loading a 0.18.3 JASP file crashes: the engine crashes when reading the emptyValues. They are apparently not upgraded in the database.

File: Desktop/data/undostack.cpp
Comment (boutinb) [reply Desktop/data/undostack.cpp]: This is a mistake, but not a serious one: the push in the undostack is not done, but it should not have occurred: probably a command leads to a refresh that leads to trigger the same command, because no check was done whether a value was really updated (as it was the mistake this commit solves). To make JASP crashing for such a mistake is quite hard I find.

**Comment by boutinb** [comment]
Also if I start JASP 0.16.3, add 'control' as empty value, load the debug.csv file and save it as JASP file, if I load this file in 0.19, then all the values of facExperim are 'control' and set as empty: the 'experiment' values are gone.

File: Desktop/data/undostack.cpp
Comment (JorisGoosen) [reply Desktop/data/undostack.cpp]: that is not actually a reason, yopu are describing the cause of the crash correctly though:
- we were not checking whether useCustom was on already which added superfluous undostuff which crashes jasp
- same for emptyValues setting

I just fixed that.

After that we do not need `_busy` for anything at all.
In fact, it would just let us miss potential problems by avoiding the crash through "_busy".
And who knows what in the future might be done with this code?
Adding extra completely unecessary checks and booleans around that does not make any sense to me...

**Comment by JorisGoosen** [comment]
> Oh loading a 0.18.3 JASP file crashes: the engine crashes when reading the emptyValues. They are apparently not upgraded in the database.

I tested this before and it worked. It would have been nice if you uploaded the file here...

Edit: Ah well it seems it broke again :p

**Comment by JorisGoosen** [comment]
And its fixed!

So if @boutinb can upload the broken file from 0.16.3 I can use that (my internet is very slow and I dont really want to waste my drivespace on 0.16.3)

In the meantime im going to try to fix internal copying of labels in JASP

**Comment by JorisGoosen** [comment]
Ok, copying labels now also works within JASP.

Ill turn on some more builds and start running some local data library tests

**Comment by JorisGoosen** [comment]
https://static.jasp-stats.org/Nightlies/macOS-arm64/JASP-nightly-emptyValuesRefactor-da7d451794d03cb921be04f6cbc4c0725c4be3a1.dmg
https://static.jasp-stats.org/Nightlies/macOS-x86_64/JASP-nightly-emptyValuesRefactor-da7d451794d03cb921be04f6cbc4c0725c4be3a1.dmg
https://static.jasp-stats.org/Nightlies/WindowsZip/JASP-nightly-emptyValuesRefactor-da7d451794d03cb921be04f6cbc4c0725c4be3a1.zip

**Comment by shun2wang** [comment]
Test feedback( on windows 11 with portable .zip):
1. load new .sav dataset will open the filter window by default?
2.  Long press<kbd>Shift</kbd>+cursor for multiple selection :selection range behavior inconsistent when selecting forward /up and backward/down. the selected range should alway 'append' but not sliding.
3. Compute new variable crash engine:
```
00:08:32.419: EngineRepresentation::handleEngineCrash():
**Engine # 0** process *crashed* and it's state is *computeColumn*.
00:08:32.423: R_HOME set to D:/Program/JASP/JASP-N~4/R
00:08:32.423: Windows PATH was changed to: 'D:/Program/JASP/JASP-nightly-emptyValuesRefactor;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin/x64'
00:08:32.508: informing engine #0 that it ought to restart
00:08:32.508: Setting new engine process to engineRepresentation Engine #0
00:08:32.546: informing engine #0 that it may resume.
00:08:32.546: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.550: Received QML warnings:
	qrc:/components/JASP/Controls/ComboBox.qml:129: TypeError: Value is undefined and could not be converted to an object

00:08:32.589: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.622: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.683: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.757: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.818: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.882: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.939: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.998: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.059: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.135: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.192: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.255: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.325: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.374: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.431: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.510: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.573: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.628: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.689: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.744: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.808: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.883: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.937: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.999: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.060: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.125: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.185: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.258: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.309: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.372: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.439: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.522: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.562: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.633: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.693: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.743: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.813: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.874: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.934: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.996: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.070: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.131: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.190: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.249: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.310: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.373: EngineRepresentation::processEngineResumedReply() for engine #0
00:08:35.374: EngineRepresentation::sendSettings()
00:08:35.374: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.435: EngineRepresentation::sendLogCfg()
00:08:35.436: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:39.647: Engine # 0 finished while in state 'computeColumn' crashing and with exitCode -1073740791!
00:08:39.647: EngineRepresentation::handleEngineCrash():
**Engine # 0** process *crashed* and it's state is *computeColumn*.
00:08:39.650: R_HOME set to D:/Program/JASP/JASP-N~4/R
00:08:39.650: Windows PATH was changed to: 'D:/Program/JASP/JASP-nightly-emptyValuesRefactor;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin/x64'
00:08:39.716: informing engine #0 that it ought to restart
00:08:39.716: Setting new engine process to engineRepresentation Engine #0
00:08:39.727: informing engine #0 that it may resume.
00:08:42.564: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:42.632: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:42.693: EngineRepresentation::processEngineResumedReply() for engine #0
00:08:42.693: EngineRepresentation::sendSettings()
00:08:42.694: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:42.754: EngineRepresentation::sendLogCfg()
00:08:42.755: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:46.194: Engine # 0 finished while in state 'computeColumn' crashing and with exitCode -1073740791!
00:08:46.194: EngineRepresentation::handleEngineCrash():
**Engine # 0** process *crashed* and it's state is *computeColumn*.
00:08:46.198: R_HOME set to D:/Program/JASP/JASP-N~4/R
00:08:46.198: Windows PATH was changed to: 'D:/Program/JASP/JASP-nightly-emptyValuesRefactor;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin/x64'
00:08:46.262: informing engine #0 that it ought to restart
00:08:46.262: Setting new engine process to engineRepresentation Engine #0
00:08:46.283: informing engine #0 that it may resume.
00:08:49.441: EngineRepresentation::processEngineResumedReply() for engine #0
00:08:49.441: EngineRepresentation::sendSettings()
00:08:49.494: EngineRepresentation::sendLogCfg()
```

File: Desktop/components/JASP/Widgets/LabelEditorWindow.qml
Comment (shun2wang) [inline Desktop/components/JASP/Widgets/LabelEditorWindow.qml]: I've always wanted to have a consistent mouse shape on the cell pointer in all table views, but that's not important:-)

Comment (JorisGoosen) [reply Desktop/components/JASP/Widgets/LabelEditorWindow.qml]: Well, this one should be only for "selecting cells" and to distinguish from the textfield/filter-checkbox.
Id prefer a better cursor than this cross but qt doesnt offer it...

**Comment by JorisGoosen** [comment]
> Test feedback( on windows 11 with portable .zip):

Thanks ^^

>     1. load new .sav dataset will open the filter window by default?

I havent seen this with my test case but maybe it does for yours?
Can you upload it?

>     2. Long pressShift+cursor for multiple selection :selection range behavior inconsistent when selecting forward /up and backward/down. the selected range should alway 'append' but not sliding.

The only inconsistency I see when using shift+cursor is that going left you first have to go through the text in the edititem.

>     3. Compute new variable crash engine:
What exactly did you do here?

**Comment by boutinb** [comment]
If a cell has a custom empty value, in a Edit mode when clicking on this cell, this removes the custom empty value.
For example, add 1 as empty value for the facFive column. Click on a cell that has 1 as value (it is displayed in grey), and go to another cell: the value 1 is removed.
I have just added a commit for this.

**Comment by shun2wang** [comment]
for 1 and 3,Sorry, in fact it can only be reproduced when I run it for the first time using .zip. After that I restart JASP and it cannot be reproduced again. I will report it separately once I find a reproducible path.

for 2, I mean row<up/down> col <left/right> will produce selection sliding behavior,see:
[screenshot_20240327.webm](https://github.com/jasp-stats/jasp-desktop/assets/10348402/e335f75c-d3cc-4742-823b-e45bde208402) press the Shift and click left/up direction will get sliding selection.

**Comment by JorisGoosen** [comment]
Ah, yeah now I see, that is definitely weird behaviour :p

**Comment by JorisGoosen** [comment]
Ok, ive fixed the selection behaviour (and improved it ^^)

**Comment by JorisGoosen** [comment]
Testing the data library revealed some bugs, that I then proceeded to fix, like:
- opening a jasp file, closing it and opening another one caused the load to fail...
- an actual missing value in the data would crash the engine on loading that column as ordinal/nominal
- a process data library file was not saved correctly. I overwrote it with the one from 0.18.3
- added a testfile with descriptives run on all columns in 0.18.3 debug.csv. Maximum statistics on in all tables (but not everything cause descriptives could not in fact handle that...)

Reloading the testfile with the emptyValuesRefactor build showed that the behaviour in Descriptives of what used to be Nominal text is hampered a bit. But im sure that requires only a light touch on the descriptives R code side

**Comment by shun2wang** [comment]
FileOperation menu has some overlap with the `▶`  , especially for language-specific strings lengths.

image

**Comment by JorisGoosen** [comment]
Ive also added a new feature to computed columns, see: https://github.com/jasp-stats/INTERNAL-jasp/issues/2487

**Comment by boutinb** [comment]
When reading a JASP file from 0.18.0 where for example 0 was set as an empty value, JASP displays the values that had 0 as empty values in the data view, but does not display the original value (0) in grey, and it does not set in the workspace that 0 is an empty value.

Sorry bad manipulation. It works: the 0 is set as empty value in the workspace. Only minor thing is that the if 0 is unset, the cells are not set back to 0: the original value (0) is lost.

**Comment by boutinb** [comment]
> Sorry bad manipulation. It works: the 0 is set as empty value in the workspace. Only minor thing is that the if 0 is unset, the cells are not set back to 0: the original value (0) is lost.

I have built it with the last commits, and now if I load a JASP file from 0.18.0, the workspace empty values are empty!

**Comment by JorisGoosen** [comment]
I only made an effort to support pre-0.18.0 versions of JASP and specifically 0.18.3.
This because there were some quite terrible bugs in those earlier 0.18.* versions and because we released them in such short order I think we can safely ignore them.

**Pull request merged**
